modules/navier_stokes/src/fvkernels/PINSFVMomentumAdvectionPorosityGradient.C

```diff
@@ -0,0 +1,73 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PINSFVMomentumAdvectionPorosityGradient.h"
+#include "PINSFVSuperficialVelocityVariable.h"
+
+registerMooseObject("NavierStokesApp", PINSFVMomentumAdvectionPorosityGradient);
+
+InputParameters
+PINSFVMomentumAdvectionPorosityGradient::validParams()
+{
+  auto params = FVElementalKernel::validParams();
+  params.addClassDescription("Porosity gradient term for the momentum equation.");
+  params.addRequiredCoupledVar("porosity", "Porosity auxiliary variable");
+
+  params.addRequiredCoupledVar("u", "The superficial velocity in the x direction.");
+  params.addCoupledVar("v", "The superficial velocity in the y direction.");
+  params.addCoupledVar("w", "The superficial velocity in the z direction.");
+
+  params.addRequiredParam<Real>("rho", "The value for the density");
+  params.declareControllable("rho");
+
+  MooseEnum momentum_component("x=0 y=1 z=2");
+  params.addRequiredParam<MooseEnum>(
+      "momentum_component",
+      momentum_component,
+      "The component of the momentum equation that this kernel applies to.");
+  params.addParam<bool>("smooth_porosity", false, "Whether the porosity has no discontinuities");
+  params.set<unsigned short>("ghost_layers") = 2;
+  return params;
+}
+
+PINSFVMomentumAdvectionPorosityGradient::PINSFVMomentumAdvectionPorosityGradient(
+    const InputParameters & params)
+  : FVElementalKernel(params),
+    _eps_var(dynamic_cast<const MooseVariableFVReal *>(getFieldVar("porosity", 0))),
+    _u(adCoupledValue("u")),
+    _v(params.isParamValid("v") ? adCoupledValue("v") : _ad_zero),
+    _w(params.isParamValid("w") ? adCoupledValue("w") : _ad_zero),
+    _rho(getParam<Real>("rho")),
+    _index(getParam<MooseEnum>("momentum_component")),
+    _smooth_porosity(getParam<bool>("smooth_porosity"))
+{
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError("PINSFV is not supported by local AD indexing. In order to use PINSFV, please run "
+             "the configure script in the root MOOSE directory with the configure option "
+             "'--with-ad-indexing-type=global'");
+#endif
+  if (!dynamic_cast<PINSFVSuperficialVelocityVariable *>(&_var))
+    mooseError("PINSFVMomentumAdvectionPorosityGradient may only be used with a superficial "
+               "velocity variable, of variable type PINSFVSuperficialVelocityVariable.");
+
+  if (!_smooth_porosity)
+    paramError(
+        "smooth_porosity",
+        "The MomentumAdvectionContinuousPorosity may only be used with a continuous porosity.");
```

Comment:
Is the intention here to remove this restriction sometime in the future? Currently the default value for `smooth_porosity` will trigger an error for a parameter that the user didn't even set.

Reply:
the intent was to force the user to set smooth_porosity.
I guess I ll make it a required param

Reply:
If in the future we find a volumetric formulation of this part of the eps * grad_P IBP that does well on porosity jump, we would remove that error for sure.

Reply:
This still feels a bit wacky to me. We are adding a required boolean parameter that only has one permissible value. But I do understand where you're coming from. You want to make sure the user understands that they should not be using this object if the porosity is not smooth. Ideally we could do a check during object construction (not technically possible) or during `initialSetup` (technically possible but O(n_face) expensive)) of whether the porosity is smooth and then error out of it is not. That would negate the need for this parameter which only has one permissible value. But acknowledging the drawbacks of these automated error checks I think what you've got here is fine